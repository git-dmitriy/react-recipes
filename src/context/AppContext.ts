import { createContext } from 'react';
import { ContextTypes } from '@/appTypes';

export const AppContext = createContext({} as ContextTypes);
