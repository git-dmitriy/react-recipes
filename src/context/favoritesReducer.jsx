import { ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE } from './types';

const handlers = {
  [ADD_TO_FAVORITE]: (state, { payload }) => {
    return {
      favorites: [...state.favorites, payload],
    };
  },
  [REMOVE_FROM_FAVORITE]: (state, { payload }) => ({
    favorites: state.favorites.filter((item) => item.idMeal !== payload.idMeal),
  }),
  DEFAULT: (state) => state,
};

export const favoriteReducer = (state, action) => {
  const handler = handlers[action.type] || handlers.DEFAULT;
  return handler(state, action);
};
