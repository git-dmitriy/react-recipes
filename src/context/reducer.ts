import {
    ADD_TO_FAVORITE,
    REMOVE_FROM_FAVORITE,
    SWITCH_THEME,
    SET_IS_LOADING,
} from '@context/contextTypes.ts';
import {StateTypes, ActionTypes} from '@/appTypes';

export const reducer = (state: StateTypes, action: ActionTypes) => {
    let favorites = null;
    switch (action.type) {
        case ADD_TO_FAVORITE:
            favorites = [...state.favorites, action.payload];
            window.localStorage.setItem('favorites', JSON.stringify(favorites));
            return {...state, favorites};

        case REMOVE_FROM_FAVORITE:
            favorites = state.favorites.filter(
                (item) => item.idMeal !== action.payload.idMeal
            );
            window.localStorage.setItem('favorites', JSON.stringify(favorites));
            return {...state, favorites};

        case SWITCH_THEME:
            window.localStorage.setItem('theme', action.payload);
            return {
                ...state,
                theme: action.payload,
            };

        case SET_IS_LOADING:
            return {
                ...state,
                isLoading: action.payload,
            };

        default:
            return state;
    }
};
