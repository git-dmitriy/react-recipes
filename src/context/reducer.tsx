import { ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE, SWITCH_THEME } from './types';
import { StateTypes, ActionTypes } from 'appTypes';

export const reducer = (state: StateTypes, action: ActionTypes) => {
  let favorites = null;
  switch (action.type) {
    case ADD_TO_FAVORITE:
      favorites = [...state.favorites, action.payload];
      window.localStorage.setItem('favorites', JSON.stringify(favorites));
      return { ...state, favorites };

    case REMOVE_FROM_FAVORITE:
      favorites = state.favorites.filter(
        (item) => item.idMeal !== action.payload.idMeal
      );
      window.localStorage.setItem('favorites', JSON.stringify(favorites));
      return { ...state, favorites };

    case SWITCH_THEME:
      window.localStorage.setItem('theme', action.payload);
      return {
        ...state,
        theme: action.payload,
      };

    default:
      return state;
  }
};
