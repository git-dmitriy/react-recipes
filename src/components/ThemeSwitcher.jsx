import { useContext } from 'react';
import { AppContext } from 'context/AppContext';
import { MdDarkMode, MdLightMode } from 'react-icons/md';

export const ThemeSwitcher = () => {
  const { state, switchTheme } = useContext(AppContext);

  const onClickHandler = () => {
    if (state.theme === 'dark') {
      switchTheme('light');
    }
    if (state.theme === 'light') {
      switchTheme('dark');
    }
  };

  return (
    <button
      className='w-8 h-7 text-3xl relative overflow-hidden sm:hover:opacity-70 sm:transition-opacity'
      onClick={onClickHandler}
    >
      <span
        className={`${
          state.theme === 'dark' ? 'rotate-180' : ''
        } absolute top-0 left-0 duration-200 transition-transform`}
      >
        <MdLightMode />
        <MdDarkMode className='rotate-180' />
      </span>
    </button>
  );
};
