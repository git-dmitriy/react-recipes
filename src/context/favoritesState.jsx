import React, { useReducer } from 'react';
import { ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE } from 'context/types';
import { FavoriteContext } from 'context/favoritesContext';
import { favoriteReducer } from 'context/favoritesReducer';

export const FavoritesState = ({ children }) => {
  const favorites = window.localStorage.getItem('favorites');
  const initialState = {
    favorites: JSON.parse(favorites) || [],
  };

  const [state, dispatch] = useReducer(favoriteReducer, initialState);

  const addToFavorites = (item) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  return (
    <FavoriteContext.Provider
      value={{
        addToFavorites,
        removeFromFavorites,
        state,
      }}
    >
      {children}
    </FavoriteContext.Provider>
  );
};
