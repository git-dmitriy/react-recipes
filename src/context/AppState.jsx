import React, { useReducer } from 'react';
import {
  ADD_TO_FAVORITE,
  SWITCH_THEME,
  REMOVE_FROM_FAVORITE,
} from 'context/types';
import { AppContext } from 'context/AppContext';
import { reducer } from 'context/reducer';

export const AppState = ({ children }) => {
  const favorites = window.localStorage.getItem('favorites');
  const initialState = {
    favorites: JSON.parse(favorites) || [],
    theme: window.localStorage.getItem('theme') || 'light',
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const addToFavorites = (item) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  const switchTheme = (mode) => {
    if (mode === 'light' || mode === 'dark') {
      dispatch({
        type: SWITCH_THEME,
        payload: mode,
      });
    }
  };

  return (
    <AppContext.Provider
      value={{
        addToFavorites,
        removeFromFavorites,
        switchTheme,
        state,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
