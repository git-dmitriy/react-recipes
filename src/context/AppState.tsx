import React, { useReducer } from 'react';
import {
  ADD_TO_FAVORITE,
  SWITCH_THEME,
  REMOVE_FROM_FAVORITE,
  SET_IS_LOADING,
} from 'context/types';
import { AppContext } from 'context/AppContext';
import { reducer } from 'context/reducer';
import { StateTypes, MealItemTypes, ContextTypes } from 'appTypes';

type P = { children: React.ReactNode };

export const AppState = ({ children }: P) => {
  const favorites = window.localStorage.getItem('favorites') || '[]';

  const initialState: StateTypes = {
    favorites: JSON.parse(favorites),
    theme: window.localStorage.getItem('theme') || 'light',
    isLoading: false,
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const addToFavorites = (item: MealItemTypes) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item: MealItemTypes) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  const switchTheme = (mode: string) => {
    if (mode === 'light' || mode === 'dark') {
      dispatch({
        type: SWITCH_THEME,
        payload: mode,
      });
    }
  };

  const setIsLoading = (status = false) => {
    dispatch({
      type: SET_IS_LOADING,
      payload: status,
    });
  };

  const value: ContextTypes = {
    addToFavorites,
    removeFromFavorites,
    switchTheme,
    setIsLoading,
    state,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
};
