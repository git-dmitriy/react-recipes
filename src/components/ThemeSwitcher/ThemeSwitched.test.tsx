import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import { AppState } from 'context/AppState';
import { ThemeSwitcher } from './ThemeSwitcher';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <AppState>
        <ThemeSwitcher />
      </AppState>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('should change theme on click', () => {
  render(
    <AppState>
      <ThemeSwitcher />
    </AppState>
  );

  const themeSwitcher = screen.getByRole('button');

  fireEvent.click(themeSwitcher);

  const icon = screen.getByText('', { selector: 'span' });

  // means change theme to dark
  expect(icon).toHaveClass('rotate-180');
});
