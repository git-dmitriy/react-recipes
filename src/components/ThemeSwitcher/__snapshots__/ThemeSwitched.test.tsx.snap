// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<button
  className="w-8 h-7 text-3xl relative overflow-hidden sm:hover:opacity-70 sm:transition-opacity"
  onClick={[Function]}
>
  <span
    className=" absolute top-0 left-0 duration-200 transition-transform"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zM2 13h2c.55 0 1-.45 1-1s-.45-1-1-1H2c-.55 0-1 .45-1 1s.45 1 1 1zm18 0h2c.55 0 1-.45 1-1s-.45-1-1-1h-2c-.55 0-1 .45-1 1s.45 1 1 1zM11 2v2c0 .55.45 1 1 1s1-.45 1-1V2c0-.55-.45-1-1-1s-1 .45-1 1zm0 18v2c0 .55.45 1 1 1s1-.45 1-1v-2c0-.55-.45-1-1-1s-1 .45-1 1zM5.99 4.58a.996.996 0 00-1.41 0 .996.996 0 000 1.41l1.06 1.06c.39.39 1.03.39 1.41 0s.39-1.03 0-1.41L5.99 4.58zm12.37 12.37a.996.996 0 00-1.41 0 .996.996 0 000 1.41l1.06 1.06c.39.39 1.03.39 1.41 0a.996.996 0 000-1.41l-1.06-1.06zm1.06-10.96a.996.996 0 000-1.41.996.996 0 00-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06zM7.05 18.36a.996.996 0 000-1.41.996.996 0 00-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06z"
      />
    </svg>
    <svg
      className="rotate-180"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 3a9 9 0 109 9c0-.46-.04-.92-.1-1.36a5.389 5.389 0 01-4.4 2.26 5.403 5.403 0 01-3.14-9.8c-.44-.06-.9-.1-1.36-.1z"
      />
    </svg>
  </span>
</button>
`;
